### ISSUE TITLE
Exclusion bug

### ISSUE THREAD

**Issue by pgrinaway** [choderalab/perses#357]
When the hybrid system is used to create a context, it sometimes throws the following exception

```
/Users/grinawap/anaconda/lib/python3.5/site-packages/simtk/openmm/openmm.py in __init__(self, *args)
   6267             a set of values for platform-specific properties. Keys are the property names.
   6268         """
-> 6269         this = _openmm.new_Context(*args)
   6270         try:
   6271             self.this.append(this)

Exception: All Forces must have identical exclusions
```

**Comment by jchodera** [comment]
Does this depend on the Platform being instantiated? I think only some platform implementations enforce this requirement.

**Comment by pgrinaway** [comment]
Oh, good call. I'll check.

**Comment by pgrinaway** [comment]
Actually, that seems to be the issue. On `Reference` and `OpenCL` (haven't tried CUDA yet) there is no issue, with or without the `ContextCache`. On `CPU` though, I get that error. Do you know why it's only enforced on some platforms? Does this signal a bug somewhere?

**Comment by jchodera** [comment]
This is either a sign that
* You are not consistently defining *exclusions* (not exceptions) across all your nonbonded `Force` terms, which is generally a bug
* You are intending to change some `NonbondedForce` parameters on the fly with `updateParametersInContext` but the exclusions start out zero and then become nonzero later on (when the kernel has already been built to exclude them)

**Comment by pgrinaway** [comment]
>You are not consistently defining exclusions (not exceptions) across all your nonbonded Force terms, which is generally a bug

Right. I just checked, and both `CustomNonbondedForce`s have the exact same exclusions.

>You are intending to change some NonbondedForce parameters on the fly with updateParametersInContext but the exclusions start out zero and then become nonzero later on (when the kernel has already been built to exclude them)

Hm, we shouldn't be doing this anywhere...

**Comment by jchodera** [comment]
> Right. I just checked, and both CustomNonbondedForces have the exact same exclusions.

They also have to match `NonbondedForce` if you have one in your system.

If you don't have a `NonbondedForce`, this sounds like an OpenMM bug. Serialize it to XML (along with `State` and `Integrator`) and post it with a simple demo script on the OpenMM issue tracker?

**Comment by pgrinaway** [comment]
>They also have to match NonbondedForce if you have one in your system.

So the `CustomNonbondedForce` _exclusions_ need to match `NonbondedForce` _exceptions_?

**Comment by pgrinaway** [comment]
I can try removing the `NonbondedForce` to check whether that was the issue.

**Comment by jchodera** [comment]
No, they need to match `NonbondedForce` exclusions, which are exceptions where `chargeProd` or `epsilon` is zero.

**Comment by pgrinaway** [comment]
Oh ok. That was obviously it. Removing the `NonbondedForce` fixed the problem. I'm going to make sure that for every exclusion, we also set an exception in the `NonbondedForce`

**Comment by jchodera** [comment]
>  I'm going to make sure that for every exclusion, we also set an exception in the NonbondedForce

Be careful with terminology. An exception with zero for `chargeProd` or `epsilon` uses `NonbondedForce.addException`, but it is an exclusion that needs to match `CustomNonbondedForce`.

**Comment by pgrinaway** [comment]
Ok. So for every exclusion, I need to call `addException` on the `NonbondedForce` to effectively add an exclusion there?

**Comment by jchodera** [comment]
Yes, if you're translating exclusions from `CustomNonbondedForce` to `NonbondedForce`.

I'm using going the other direction, where I am recoding a `NonbondedForce` into `CustomNonbondedForce` (which must hold the exclusions) and `CustomBondForce` (which holds the exceptions).

**Comment by pgrinaway** [comment]
Ok, looking into this, I don't think this is a "bug" in my code exactly--the exclusions don't match on purpose. I'm using "exclusions" (exceptions that are zeroed out) in the `NonbondedForce` to avoid computing interactions between atoms covered by the `CustomNonbondedForce`s. However, in the `CustomNonbondedForce`, the double-counting is avoided by using interaction groups (which don't exist in the `NonbondedForce`)

I can resolve this by changing from using interaction groups to exclusions (though it seemed the interaction group approach created much cleaner code). However, since it doesn't seem to cause an issue on Reference, OpenCL, or CUDA, I labeled this issue low priority.

**Comment by jchodera** [comment]
> Ok, looking into this, I don't think this is a "bug" in my code exactly--the exclusions don't match on purpose. 

That's a bug, since it doesn't work on all platforms. :)

**Comment by jchodera** [comment]
> the double-counting is avoided by using interaction groups (which don't exist in the NonbondedForce)

As long as you're not using PME, you can re-code anything as a `CustomNonbondedForce` so you *can* use interaction groups.

Do remember that cutoffs aren't used in their calculation, though, so this may be inefficient.

**Comment by pgrinaway** [comment]
>That's a bug, since it doesn't work on all platforms. :)

Sure, ok. But I think getting CPU platform to work is relatively low priority, since we can use OpenCL with CPU if we really want. I think I should probably raise an issue about documenting this aspect of OpenMM--it's highly non-intuitive.

>As long as you're not using PME, you can re-code anything as a CustomNonbondedForce so you can use interaction groups. Do remember that cutoffs aren't used in their calculation, though, so this may be inefficient.

Right--I recall you telling me this, so I tried to minimize my usage of both `CustomNonbondedForce` and interaction groups.

**Comment by jchodera** [comment]
> I think I should probably raise an issue about documenting this aspect of OpenMM--it's highly non-intuitive.

Agreed---it's a poorly documented limitation of the platform.

> Right--I recall you telling me this, so I tried to minimize my usage of both CustomNonbondedForce and interaction groups.

We could get the efficiency thing fixed if we had a good use case.

**Comment by pgrinaway** [comment]
Ok, I've moved this issue to the "to discuss" column because changing this behavior will require substantial work, and it seems relatively low priority.

**Comment by jchodera** [comment]
Maybe that's something I should address when I move this to openmmtools.alchemy rather than something you should address now.

**Comment by pgrinaway** [comment]
SGTM

**Comment by Derivolate** [comment]
> This is either a sign that
> 
>     * You are not consistently defining _exclusions_ (not exceptions) across all your nonbonded `Force` terms, which is generally a bug
> 

I am building an FEP simulation for EVB (Empirical Valence Bond) in OpenMM in which I am interpolating between two forcefields. These two forcefields have different exceptions/exclusions and for any interpolated state that has contributions from both these forcefields, I would need to construct a forcefield that has 2 non-bonded forces with different exceptions/exclusions. The workaround for this would be to construct a custom nonbonded force that has the proper interpolated contributions, but merging the exceptions/exclusions is a nightmare.

**Comment by ajerschow** [comment]
just confirming that switching to CUDA solved the problem for me with a custom NB force where I had two exclusions and an interaction group.  I realize it's not an ideal solution for every case.

**Comment by gosthak** [comment]
I have kinda the same problem. I need to set the interaction using BondedForce between two specific pair of atoms as an exception in order to modify the interaction parameters during the simulation. But I have customNonbondedForce acting between them as well, and this custom interaction I need to exclude.

Can you tell me how to break this impasse?
